File: src/LinkedList/DoubleLinkedList/DoubleLinkedList.java
package LinkedList.DoubleLinkedList;
/*
 * @author Ionésio Junior
 * */

/*
 * Double Linked List in Iterative implementation
 * */
public class DoubleLinkedList<T> implements DoubleLinkedListInterface<T> {

	private DoubleNode<T> head;
	private DoubleNode<T> tail;
	
	public DoubleLinkedList() {
		this.head = new DoubleNode<T>();
		this.tail = new DoubleNode<T>();
		this.head.setNext(this.tail);
		this.tail.setPrevious(this.head);
	}
	
	@Override
	public boolean isEmpty() {
		return this.head.isNIL();
	}

	@Override
	public int size() {
		int size = 0;
		DoubleNode<T> aux = this.head;
		while(!aux.isNIL()){
			aux = aux.getNext();
			size++;
		}
		return size;
	}

	@Override
	public T search(T element) {
		if(element != null){
			DoubleNode<T> aux = this.head;
			while(!aux.isNIL()){
				if(aux.getData().equals(element)){
					return aux.getData();
				}
				aux = aux.getNext();
			}
			return null;
		}else{
			return null;
		}
	}

	@Override
	public void insert(T element) {
		if(element != null){
			if(this.isEmpty()){
				this.insertFirst(element);
			}else{
				DoubleNode<T> node = new DoubleNode<T>(element,this.tail,new DoubleNode<T>());
				this.tail.setNext(node);
				this.tail = node;
			}
		}
	}

	@Override
	public void remove(T element) {
		if(element != null && !this.isEmpty()){
			if(this.head.getData().equals(element)){
				this.removeFirst();
			}else{
				DoubleNode<T> aux = this.head;
				while(!aux.isNIL()){
					if(aux.getData().equals(element)){
						aux.getPrevious().setNext(aux.getNext());
						aux.getNext().setPrevious(aux.getPrevious());
						if(aux.getNext().isNIL()){
							this.tail = aux.getPrevious();
						}
						break;
					}
					aux = aux.getNext();
				}
			}
		}
	}

	@Override
	public T[] toArray() {
		T[] result = (T[]) new Object[this.size()];
		DoubleNode<T> aux = this.head;
		int index = 0;
		while(!aux.isNIL()){
			result[index++] = aux.getData();
			aux = aux.getNext();
		}
		return result;
	}

	@Override
	public void insertFirst(T element) {
		if(element != null){
			if(this.isEmpty()){
				DoubleNode<T> newHead = new DoubleNode<T>(element,new DoubleNode<T>(),new DoubleNode<T>());
				this.head = newHead;
				this.tail = newHead;
			}else{
				DoubleNode<T> newHead = new DoubleNode<T>(element,new DoubleNode<T>(),this.head);
				this.head.setPrevious(newHead);
				this.head = newHead;
			}
		}
	}

	@Override
	public void removeFirst() {
		if(this.isEmpty())
			return;
		else{
			this.head.getNext().setPrevious(new DoubleNode<T>());
			if(this.size() == 1){
				this.tail = this.head.getNext();
			}
			this.head = this.head.getNext();
		}
	}

	@Override
	public void removeLast() {
		if(!this.isEmpty()){
			this.tail.getPrevious().setNext(new DoubleNode<T>());
			if(this.size() == 1){
				this.head = this.tail.getPrevious();
			}
			this.tail = this.tail.getPrevious();
		}
	}
}

File: Python/LinkedList/DoubleLinkedList/DoubleLinkedList.py
from DoubleNode import DoubleNode

class DoubleLinkedList():
	
	def __init__(self):
		self.__head = DoubleNode()
		self.__tail = DoubleNode()
		self.__head.setNext(self.__tail)
		self.__tail.setPrevious(self.__head)
	
	
	def isEmpty(self):
		return (self.__head.isEmpty())
	
	
	def size(self):
		size = 0
		aux = self.__head
		while(not(aux.isEmpty())):
			aux = aux.getNext()
			size = size + 1
		return size
	
	
	def search(self, element):
		if(element != None):
			aux = self.__head
			while(not(aux.isEmpty())):
				if(aux.getData() == element):
					return aux.getData()
				aux = aux.getNext()
	
	def insert(self,element):
		if(element != None):
			if(self.isEmpty()):
				self.insertFirst(element)
			else:
				node = DoubleNode(element,self.__tail,DoubleNode())
				self.__tail.setNext(node)
				self.__tail = node
	
	def remove(self,element):
		if(element != None and not(self.isEmpty())):
			if(self.__head.getData() == element):
				self.removeFirst()
			else:
				aux = self.__head
				while(not(aux.isEmpty())):
					if(aux.getData() == element):
						aux.getPrevious().setNext(aux.getNext())
						aux.getNext().setPrevious(aux.getPrevious())
						if(aux.getNext().isEmpty()):
							self.__tail = aux.getPrevious()
						break
					aux = aux.getNext()
	
	
	def toArray(self):
		result = []
		aux = self.__head
		while(not(aux.isEmpty())):
			result.append(aux.getData())
			aux = aux.getNext()
		return result
	
	
	def insertFirst(self,element):
		if(element != None):
			if(self.isEmpty()):
				newHead = DoubleNode(element, DoubleNode(),DoubleNode())
				self.__head = newHead
				self.__tail = newHead
			else:
				newHead = DoubleNode(element,DoubleNode(),self.__head())
				self.__head.setPrevious(newHead)
				self.__head = newHead
	
	
	def removeFirst(self):
		if(self.isEmpty()):
			return
		else:
			self.__head.getNext().setPrevious(DoubleNode())
			if(self.size() == 1):
				self.__tail = self.__head.getNext()
			self.__head = self.__head.getNext()
	
	
	def removeLast(self):
		if(self.isEmpty()):
			self.__tail.getPrevious().setNext(DoubleNode())
			if(self.size() == 1):
				self.__head = self.tail.getPrevious()
			self.__tail = self.tail.getPrevious()
	
		
